core-bundle/src/Security/Authentication/RememberMe/ExpiringTokenBasedRememberMeServices.php

```diff
@@ -0,0 +1,253 @@
+            && 2 === \count($parts = $this->decodeCookie($cookie))
+        ) {
+            [$series] = $parts;
+            $this->connection->delete('tl_remember_me', ['series' => $series]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processAutoLoginCookie(array $cookieParts, Request $request): UserInterface
+    {
+        if (2 !== \count($cookieParts)) {
+            throw new AuthenticationException('The cookie is invalid.');
+        }
+
+        [$series, $tokenValue] = $cookieParts;
+
+        try {
+            $this->connection->exec('LOCK TABLES tl_remember_me WRITE');
+
+            $rows = $this->loadTokens($series);
+
+            $currentToken = null === $rows[0]->expires ? $rows[0] : null;
+            $matchedToken = $this->findValidToken($rows, $tokenValue);
+
+            $this->cleanupExpiredTokens();
+
+            if (null === $currentToken || $currentToken->value === $matchedToken->value || null !== $matchedToken->expires) {
+                $tokenValue = $this->migrateToken($matchedToken);
+            } else {
+                $tokenValue = $currentToken->value;
+            }
+        } catch (\Exception $e) {
+            if ($e instanceof AuthenticationException) {
+                throw $e;
+            }
+
+            throw new AuthenticationException('Rememberme services failed.', 0, $e);
+        } finally {
+            $this->connection->exec('UNLOCK TABLES');
+        }
+
+        $request->attributes->set(self::COOKIE_ATTR_NAME, $this->createCookie($request, $series, $tokenValue));
+
+        return $this->getUserProvider($matchedToken->class)->loadUserByUsername($matchedToken->username);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function onLoginSuccess(Request $request, Response $response, TokenInterface $token): void
+    {
+        $series = base64_encode(random_bytes(64));
+        $tokenValue = base64_encode(random_bytes(64));
+
+        $this->insertToken(
+            \get_class($user = $token->getUser()),
+            $user->getUsername(),
+            $series,
+            $tokenValue
+        );
+
+        $response->headers->setCookie($this->createCookie($request, $series, $tokenValue));
+    }
+
+    private function loadTokens(string $series): array
+    {
+        try {
+            $rows = $this->connection->executeQuery(
+                'SELECT * FROM tl_remember_me WHERE series=:series AND (expires IS NULL OR expires>=:expires) ORDER BY expires IS NULL DESC',
+                [
+                    'series' => $series,
+                    'expires' => new \DateTime(),
+                ],
+                [
+                    'series' => \PDO::PARAM_STR,
+                    'expires' => DoctrineType::DATETIME,
+                ]
+            )->fetchAll(\PDO::FETCH_OBJ);
+        } catch (\Exception $e) {
+            $rows = [];
+        }
+
+        if (empty($rows)) {
```

Comment:
I would use `0 === \count($rows)`. It's type safe and faster than `empty()`.

Reply:
That cannot be correct, because `empty()` is a language construct and thus should be faster than any function.

Reply:
An array is a language construct too and it saves the number of items internally.
See for yourself: https://3v4l.org/TV5q2/perf#output and https://3v4l.org/ldDUr/perf#output

Reply:
As soon as the array is not empty, `empty()` has the same performance on PHP 7.3 but uses significantly less memory… https://3v4l.org/vF96k/perf#output vs https://3v4l.org/ZHlvf/perf#output

Reply:
The differences in performance are irrelevant anyway. They are way too small. So that put aside, `count()` is the way to go for type safety 🥁

Reply:
I'm fine with using `count()` then.

> is the way to go for type safety

Only if you accept arrays **and** `Countable` objects.

Reply:
https://github.com/kalessil/phpinspectionsea/blob/master/docs/types-compatibility.md#empty-usage